### DISCUSSION TITLE
Filtered list

### DISCUSSION THREAD

**Discussion by alekseyshaklunov** [macgile/DataGridFilter#181] [Q&A]
Hello! I apologize in advance if the question is silly. How can I get a filtered list? For example, using Binding or in some other way?

**Comment by macgile** [comment]
Hi,
The demo application in the repository is a good start.
best regard

**Reply by alekseyshaklunov** [reply]
Before asking my question, I reviewed DemoApp.Net7.0 (SharedModelView) and the description on Code Project, but I didn’t find any way to obtain filtered data. The FilteredList remains in its original state after filtering. I’m not a professional programmer (and not a programmer at all), so if this is only possible by modifying the FilterDataGrid.Net project, I currently don’t have the skills for that and will look for another solution. Could you please let me know if it’s possible to obtain a filtered collection without modifying the FilterDataGrid.Net project? Thank you!

**Comment by macgile** [comment]
Yes it is possible but it requires to do some coding.
Look for examples of filtering with LINQ (Language-Integrated Query)
Good luck

**Reply by alekseyshaklunov** [reply]
I figured out how to get the filtered collection. Thanks. However, now another problem has arisen, hopefully the last one. I need to dynamically create columns in the code behind after the user clicks a button, i.e., when the WPF window is already created, and the DataContext is set. But with this approach, the filter lists in the WPF form remain empty – only 'Select all' and 'Blank' are displayed. If I create the columns after filling them with data, then the WPF form lacks the filtering buttons, even though IsColumnFiltered = true. Is there perhaps a way to refresh the filter list that I'm not seeing? I tried fully creating the FilterDataGrid in the code behind, but the result is the same. Code is below. Thank you!

```
_filterDataGrid = new FilterDataGrid.FilterDataGrid();
_filterDataGrid.AutoGenerateColumns = false;

var binding = new Binding("Nodes")
{
    Source = _viewModel,
    UpdateSourceTrigger = UpdateSourceTrigger.PropertyChanged
};
BindingOperations.SetBinding(_filterDataGrid, ItemsControl.ItemsSourceProperty, binding);

foreach (string property in _viewModel.Columns)
{
    FilterDataGrid.DataGridTextColumn column = new FilterDataGrid.DataGridTextColumn()
    {
        Header = property,
        Binding = new Binding($"Properties[{property}]"),
        IsColumnFiltered = true
    };
    _filterDataGrid.Columns.Add(column);
}

BaseGrid.Children.Add(_filterDataGrid);
```

**Comment by macgile** [comment]
Try
```
var binding = new Binding
{
    Source = _viewModel, // The data source collection
    UpdateSourceTrigger = UpdateSourceTrigger.PropertyChanged
};
BindingOperations.SetBinding(filterDataGrid, ItemsControl.ItemsSourceProperty , binding);
```

```
var column = new FilterDataGrid.DataGridTextColumn()
{
   // property  =  Name of field
    Binding = new Binding(property),
    IsColumnFiltered = true,
    FieldName =property
};
filterDataGrid.Columns.Add(column);

```

**Reply by alekseyshaklunov** [reply]
Unfortunately, this did not help. The Node class has a property Dictionary<string, object> Properties, where the key is the column header and the value is the cell value. Here’s what I tried:
1. The code from the previous message, including your modifications.
2. Conversion to ExpandoObject.
3. Using an indexer.
4. Newtonsoft.Json serialization/deserialization.
5. Created a separate WPF window to execute everything before setting the DataContext.
6. Created FilterDataGrid both in XAML and in code behind.

The result is always the same - the data in FilterDataGrid is displayed correctly, but the filters:
1. If I create the columns before filling List<Node>, the filters are empty (only 'Select all' and 'Blank' are displayed).
2. If I create the columns after filling List<Node>, there are not even filter buttons.
I really need help, at least in which direction the solution might be.
